crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1101,6 +1118,67 @@ impl<'db> TypeInferenceBuilder<'db> {
             self.infer_definition(parameter);
         }
         self.infer_body(&function.body);
+
+        if let Some(declared_ty) = function
+            .returns
+            .as_deref()
+            .map(|ret| self.file_expression_type(ret))
+        {
+            fn is_suite_empty(suite: &[ast::Stmt]) -> bool {
+                let mut saw_ellipsis = false;
+                for statement in suite {
+                    if let Some(expr) = statement.as_expr_stmt() {
+                        if expr.value.is_ellipsis_literal_expr() {
+                            saw_ellipsis = true;
+                        } else if expr.value.is_string_literal_expr() {
+                            // docstring
+                        } else {
+                            return false;
+                        }
+                    }
+                }
+                saw_ellipsis
+            }
+            if is_suite_empty(&function.body) {
+                return;
+            }
+            let inferred_ty = if self.types.return_types_and_ranges.is_empty() {
+                KnownClass::NoneType.to_instance(self.db())
+            } else {
+                UnionType::from_elements(
+                    self.db(),
+                    self.types
+                        .return_types_and_ranges
+                        .iter()
+                        .map(|ty_range| ty_range.ty),
+                )
```

Comment:
I don't think we should construct a union type for a single initial assignability check. Assignability checking for a union generally means iterating all its member types anyway, and this way we also pay the cost to construct the union type. Why not iterate all possible return types and ranges (after adding "end-of-body `None`" to the list) and check each one for assignability to the annotated return type (that is, what we end up doing here in the error case anyway), without first doing a union assignability check?